bpf-sys/src/type_gen.rs

```diff
@@ -0,0 +1,253 @@
+syntax. So macro constants can not be generated from vmlinux image. But
+`type_gen` provides common macro constants by including some C header files in
+system.
+*/
+
+use super::{
+    __va_list_tag, btf, btf__free, btf__get_nr_types, btf__name_by_offset, btf__parse_elf,
+    btf__type_by_id, btf_dump, btf_dump__dump_type, btf_dump__free, btf_dump__new, btf_dump_opts,
+    libbpf_find_kernel_btf, vfprintf,
+};
+use libc::{self, c_char, c_void, FILE};
+use regex::RegexSet;
+use std::env;
+use std::ffi::{CStr, CString};
+use std::io;
+use std::mem::MaybeUninit;
+use std::path::Path;
+use std::path::PathBuf;
+use std::ptr;
+const ENV_VMLINUX_PATH: &'static str = "REDBPF_VMLINUX";
+
+// only used for RAII
+struct StreamWrapper(*mut FILE);
+impl Drop for StreamWrapper {
+    fn drop(&mut self) {
+        unsafe {
+            libc::fclose(self.0);
+        }
+    }
+}
+
+// only used for RAII
+struct BtfDumpWrapper(*mut btf_dump);
+impl Drop for BtfDumpWrapper {
+    fn drop(&mut self) {
+        unsafe {
+            btf_dump__free(self.0);
+        }
+    }
+}
+
+/// An error that occurred during parsing vmlinux and generating C source code
+/// from BTF.
+#[derive(Debug)]
+pub enum TypeGenError {
+    /// error on parsing vmlinux
+    VmlinuxParsingError,
+    /// vmlinux not found
+    VmlinuxNotFound,
+    /// path contains invalid utf-8
+    InvalidPath,
+    /// IO error
+    IO(io::Error),
+    /// invalid regex
+    RegexError,
+    DumpError,
+}
+
+type Result<T> = std::result::Result<T, TypeGenError>;
+
+/// Load vmlinux from file and generate `vmlinux.h`
+pub struct VmlinuxBtfDump {
+    whitelist: Option<Vec<String>>,
+    btfptr: *mut btf,
+}
+
+impl VmlinuxBtfDump {
+    /// Probe few well-known locations for vmlinux kernel image and try to load
+    /// BTF data out of it
+    pub fn with_system_default() -> Result<Self> {
+        let btfptr = unsafe { libbpf_find_kernel_btf() };
+        if (btfptr as isize) < 0 {
+            return Err(TypeGenError::VmlinuxNotFound);
+        }
+
+        Ok(VmlinuxBtfDump {
+            whitelist: None,
+            btfptr,
+        })
+    }
+
+    /// Try to load BTF data out of the given ELF file
+    pub fn with_elf_file(elf_file: impl AsRef<Path>) -> Result<Self> {
+        if !elf_file.as_ref().exists() {
+            return Err(TypeGenError::VmlinuxNotFound);
+        }
+
+        let elf_str = elf_file
+            .as_ref()
+            .to_str()
+            .ok_or(TypeGenError::InvalidPath)?;
+
+        let cvmlinux_path = CString::new(elf_str).unwrap();
+        let btfptr = unsafe { btf__parse_elf(cvmlinux_path.as_ptr(), ptr::null_mut()) };
+        if (btfptr as isize) < 0 {
+            return Err(TypeGenError::VmlinuxParsingError);
+        }
+
+        Ok(VmlinuxBtfDump {
+            whitelist: None,
+            btfptr,
+        })
+    }
+
+    /// Add regex `pattern` into whitelist of BTF types
+    pub fn whitelist(mut self, pattern: &str) -> Self {
```

Comment:
Same here, would be nicer with `allowlist` as terminology.

Reply:
Good. I will rename this.